controllers/provider_controller.go

```diff
@@ -61,26 +62,36 @@ func (r *ProviderReconciler) Reconcile(ctx context.Context, req ctrl.Request) (c
 		return ctrl.Result{}, err
 	}
 
-	if _, err := providercred.GetProviderCredentials(ctx, r.Client, &provider, provider.Spec.Region); err != nil {
-		provider.Status.State = types.ProviderIsNotReady
-		provider.Status.Message = fmt.Sprintf("%s: %s", errGetCredentials, err.Error())
-		klog.ErrorS(err, errGetCredentials, "Provider", req.NamespacedName)
-		if updateErr := r.Status().Update(ctx, &provider); updateErr != nil {
-			klog.ErrorS(updateErr, errSettingStatus, "Provider", req.NamespacedName)
-			return ctrl.Result{}, errors.Wrap(updateErr, errSettingStatus)
+	err := func() error {
+		switch provider.Spec.Credentials.Source {
+		case crossplanetypes.CredentialsSourceInjectedIdentity:
+			break
+		case crossplanetypes.CredentialsSourceSecret:
+			_, err := providercred.GetProviderCredentials(ctx, r.Client, &provider, provider.Spec.Region)
+			return err
+		default:
+			return errors.Errorf("unsupported credentials source: %s", provider.Spec.Credentials.Source)
 		}
-		return ctrl.Result{}, errors.Wrap(err, errGetCredentials)
-	}
 
-	provider.Status = terraformv1beta1.ProviderStatus{
-		State: types.ProviderIsReady,
+		return nil
+	}()
+	if err != nil {
+		klog.ErrorS(err, errGetCredentials, "Provider", req.NamespacedName)
+
+		provider.Status.Message = fmt.Sprintf("%s: %s", errGetCredentials, err.Error())
+		provider.Status = terraformv1beta1.ProviderStatus{State: types.ProviderIsNotReady}
+	} else {
+		provider.Status.Message = "Provider ready"
+		provider.Status = terraformv1beta1.ProviderStatus{State: types.ProviderIsReady}
 	}
+
 	if updateErr := r.Status().Update(ctx, &provider); updateErr != nil {
 		klog.ErrorS(updateErr, errSettingStatus, "Provider", req.NamespacedName)
+
 		return ctrl.Result{}, errors.Wrap(updateErr, errSettingStatus)
 	}
 
-	return ctrl.Result{}, nil
+	return ctrl.Result{}, err
```

Comment:
Is the err https://github.com/oam-dev/terraform-controller/pull/295/files#diff-8d523b8a346b8e0143bd91409158007d7870cee18c5facd66f70b2c94c8c2f1fR65?

Reply:
Indeed .. though admitted not the best way of doing it .. i wasn't sure where else as you always want the Status() to be updated.

Reply:
Then we had better not return err here. The rest are LGTM.